doc/man/bxl2txt.1
.\" pcb-rnd - manual
.\" Copyright (C) 2016 Tibor 'Igor2' Palinkas
.\" 
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\" 
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
.\" GNU General Public License for more details.
.\" 
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write to the Free Software Foundation, Inc.,
.\" 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
.\" 
.\" Contact: pcb-rnd-man[removethis]@igor2.repo.hu
.TH bxl2txt 1 2020-04-07 "" "pcb-rnd manual"
.SH NAME
bxl2txt - decode a binary BXL file and dump the plain text content
.SH SYNPOSIS
.nf
.sp
\fBbxl2txt < \fIinfile.bxl\fB > \fIoutfile.txt\fB
.fi
.SH DECSRIPTION

.BR bxl2txt
reads and decodes a binary bxl file in its standard input and prints the plain text content on its standard output.


Changelog
pcb-rnd 2.2.0 (r29205)
~~~~~~~~~~~~~~~~~~~~~~

	[act_draw]
		-Add: action for layer object duplication

	[act_read]
		-Add: ReadGroup(): action for querying group fields

	[asm]
		-Fix: template for x and y didn't append the coordinate values to the right string

	[autocrop]
		-Fix: both board size change and data move should be undoable

	[bom]
		-Cleanup: remove local string list implementation in favor of common vts0
		-Cleanup: name space conventions
		-Cleanup: no need to cast malloc() return value, the code is not C++
		-Cleanup: code formatting & whitespace
		-Fix: obsolete comment on what the plugin does

	[cam]
		-Fix: default cam jobs: gerber assy exports always export both the virtual and the doc layer for compatibility with old and new stackups
		-Fix: negative layer group offset sign bug (should count _back_ from the last group)
		-Add: doc: mention [output options] in the syntax comment
		-Add: executing a cam export script should be a single export session

	[core]
		-Cleanup: remove dolists.h and make action table registration macros more explicit to simplify the API
		-Cleanup: attribute change callback should always happen, including when the attribute is free'd so administration depending on the callback will always work
		-Cleanup: attribute memory leak at object free
		-Cleanup: centralize/move subc-terminal-hash administration into attrib post-change handler, so it can not miss any term change
		-Cleanup: actions: get hidlib through argv0 instead if using global var PCB
		-Cleanup: change order of functions in main.c so path initialziation is on top, close to path var declarations
		-Cleanup: simplify the exec_prefix code by not re-implementing strchr()
		-Cleanup: de-doxyfy the exec_prefix comment
		-Cleanup: remove use of appearance/color/crosshair - redundant with appearance/color/cross
		-Cleanup: const correctness in pcb_poly_contour_copy(), pcb_poly_contour_new(), cntrbox_adjust(), pcb_poly_contour_inside()
		-Cleanup: footprint API: take a conf list instead of a ':' separated lsit of strings for footprint search path
		-Cleanup: centralize pcb aux unit vector size as more and more parts of the code start to depend on it
		-Fix: do draw arcs with 0 thickness, only omit negative thickness objects
		-Fix: minimum thickness should be minimum size and maximum thickness should be maximum coord - don't constraint thickness beyond our number representation
		-Fix: allow arc radius to go down to 0 and text to scale down to 1% - don't constraint it with a random value when nothing breaks
		-Fix: do update subc->terminals hash if a terminal is removed or added back by undo
		-Fix: still draw "cleared out of existence" polygon's as-drawn contour if it was requested
		-Fix: objects that are being free'd need to be removed from their rtree
		-Fix: attribute free should be done in 2 steps (notify then free), to avoid cached pointers dangling
		-Fix: run-time concat of config paths - don't assume all parts are compile-time-known (helps win32)
		-Fix: pcb_crosshair.snapped_pstk was never set back to NULL after leaving the snapped padstack
		-Fix: Cursor() steps really in grid unit when addressed by grid in last arg and is not snapped
		-Fix: Cursor(): when leaving the snapped padstack, don't go much further than really needed
		-Fix: file-action: new pcb action legitimately uses the PCB variable directly, but it needs to notify the GUI about the new PCB as soon as possible
		-Fix: default boards: group name for top and bottom assy should have _ for separator, not -, for consistency
		-Fix: set color before drawing marks.
		-Fix: do not set crosshair color if nothing is drawn from the crosshair
		-Fix: polygon support: pline's head vnode is a pointer, not a local struct - this adds one extra allocation per pline, but removes the special casing for head, which is required for the 'exclude head' bugfix later
		-Fix: pcb_poly_vertex_exclude() can be called with the pline the node is deleted from; if the node is the head, the pline's head is bumped to the next node to avoid dangling head
		-Fix: thermal/padstack-shape poly render corner case: in gcode, the hole and slot shape does need to be subtracted on mech layers
		-Fix: padstack x-mirroring: also have to reverse the order of polygon points to keep the normal vectors in the right direction (wihout this, shrinking/growing will do the opposite)
		-Fix: fp fopen: return the first footprint found instead of returning NULL if the last search path didn't have it
		-Fix: poly draw helper: when drawing optimization (single line instead of polygon too small), set line cap and width
		-Fix: move-to-current-layer: make sure the original (sometimes bound) layer (parent of the object) is passed as original layer instead of the resolved board layer so the undo would put the object back to its right layer
		-Fix: move-to-layer undo: do not assume the other layer is in PCB, it can be a subc layer as well
		-Fix: NewGroup(): uninitialized attr argument caused segfaults in some cases
		-Fix: allow setting clearance value through the Change infrastructure to smaller than DRC minimum (pcb-rnd is not a nanny, it should just do what the user ordered)
		-Fix: pcb_lineop_move_point_with_route() shall return the original line object instead of NULL (extobj will need it)
		-Fix: route code should copy attributes of lines and arcs being split
		-Fix: line endpoint move with the route code does not swap endpoints
		-Fix: invalid memory handling on routing style caused by emitting the event before the correct pcb is made global
		-Fix: inhibit redraws while creating a new board
		-Fix: moving arc endpoint did not set return value properly, caused uninitialized pointer return
		-Fix: when undoing flags, re-clip on flag changes that affect poly clipping (the CLEAR* flags)
		-Fix: make low level calls of board size change undoable (will fix non-undoable autocrop)

util/bxl2txt/bxl2txt.c
/*
 *                            COPYRIGHT
 *
 *  pcb-rnd, interactive printed circuit board design
 *
 *  BXL IO plugin - file format read, parser
 *  pcb-rnd Copyright (C) 2020 Tibor 'Igor2' Palinkas
 *  (Supported by NLnet NGI0 PET Fund in 2020)
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 *  Contact:
 *    Project page: http://repo.hu/projects/pcb-rnd
 *    lead developer: http://repo.hu/projects/pcb-rnd/contact.html
 *    mailing list: pcb-rnd (at) list.repo.hu (send "subscribe")
 */

/* test program: read bxl on stdin, unpack to stdout */

#include <stdio.h>
#include "bxl_decode.h"

int main()
{
	hdecode_t ctx;
	int inch;

	pcb_bxl_decode_init(&ctx);

	while((inch = fgetc(stdin)) != EOF) {
		int n, len;
		len = pcb_bxl_decode_char(&ctx, inch);
/*		for(n = 0; n < len; n++)
			printf("%c %d\n", ctx.out[n], ctx.out[n]);*/
		for(n = 0; n < len; n++)
			printf("%c", ctx.out[n]);
	}

/*fprintf(stderr, "plain len=%lu %lu\n", ctx.plain_len, ctx.ipos);*/
/*fprintf(stderr, "pool used: %d\n", ctx.tree->pool_used);*/

	return 0;
}
